contracts/src/v0.8/dev/transmission/4337/Paymaster.sol

```
@@ -0,0 +1,118 @@
+    if (msg.sender != address(LINK)) {
+      revert OnlyCallableFromLink();
+    }
+    if (_data.length != 32) {
+      revert InvalidCalldata();
+    }
+
+    address subscription = abi.decode(_data, (address));
+    subscriptions[subscription] += _amount;
+  }
+
+  function validatePaymasterUserOp(
+    UserOperation calldata userOp,
+    bytes32 userOpHash,
+    uint256 maxCost
+  ) external returns (bytes memory context, uint256 validationData) {
+    if (userOpHashMapping[userOpHash]) {
+      revert UserOperationAlreadyTried(userOpHash);
+    }
+
+    uint256 extraCostJuels = handleExtraCostJuels(userOp);
+    uint256 costJuels = getCostJuels(maxCost) + extraCostJuels;
+    if (subscriptions[userOp.sender] < costJuels) {
+      revert InsufficientFunds(costJuels, subscriptions[userOp.sender]);
+    }
+
+    userOpHashMapping[userOpHash] = true;
+    return (abi.encode(userOp.sender, extraCostJuels), _packValidationData(false, 0, 0)); // success
+  }
+
+  /// @dev Calculates any extra LINK cost for the user operation, based on the funding type passed to the
+  /// @dev paymaster. Handles funding the LINK token funding described in the user operation.
+  /// TODO: add logic for subscription top-up.
+  function handleExtraCostJuels(UserOperation calldata userOp) internal returns (uint256 extraCost) {
+    if (userOp.paymasterAndData.length == 20) {
+      return 0; // no extra data, stop here
+    }
+
+    uint8 paymentType = uint8(userOp.paymasterAndData[20]);
+
+    // For direct funding, use top-up logic.
+    if (paymentType == uint8(SCALibrary.LinkPaymentType.DIRECT_FUNDING)) {
+      SCALibrary.DirectFundingData memory directFundingData = abi.decode(
+        userOp.paymasterAndData[21:],
+        (SCALibrary.DirectFundingData)
+      );
+      if (
+        directFundingData.topupThreshold != 0 &&
+        LINK.balanceOf(directFundingData.recipient) < directFundingData.topupThreshold
```

maybe better to revert in this scenario so that we don't silently proceed when insufficient balance?

---

we could use `SafeERC20.safeTransfer` for error handling

---

This simply checks if the end-contract needs a top-up, it doesn't check for an insufficient balance

---

Ah I see. Should we still handle `transfer` failures though? `transfer()` can fail if paymaster isn't topped up right?

---

Never mind. Looking at LINK token implementation, it never returns false for `transfer`